distsql: adds support for distributed bulk ops in multi-tenant

rharding6373 opened the pull request:
distsql: adds support for distributed bulk ops in multi-tenant
    
Previously, distsql did not support distributed queries in multi-tenant
environments. Queries were always planned using a single SQL pod, even
if using multiple pods could yield better parallelism.
    
This change adds an interface inside distsql that allows different
implementations for functions that differ between single- and
multi-tenant. Currently, it only has `SetupAllNodesPlanning`, which for
single-tenant remains the same, but in multi-tenant returns all
available SQL pods at the time it is invoked.
    
This PR also fixes some places where SQL pods need to communicate with
other SQL pods using the PodNodeDialer instead of the NodeDialer that
handles SQL <-> KV communication in multi-tenant mode.

Fixes #74548

References #47900
    
Release note (sql change): Adds support for distributed import queries in
multi-tenant environments, which allows import queries to have improved
parallelism by utilizing all available SQL pods in the tenant.

importccl, backupccl: add import and restore tenant tests with multiple SQL pods

These tests validate that imports and restores can be
distributed across multiple SQL pods in a single tenant.

Release note: None

rharding6373 committed:
distsql: adds support for distributed bulk ops in multi-tenant

Previously, distsql did not support distributed queries in multi-tenant
environments. Queries were always planned using a single SQL pod, even
if using multiple pods could yield better parallelism.

This change adds an interface inside distsql that allows different
implementations for functions that differ between single- and
multi-tenant. Currently, it only has `SetupAllNodesPlanning`, which for
single-tenant remains the same, but in multi-tenant returns all
available SQL pods at the time it is invoked.

This PR also fixes some places where SQL pods need to communicate with
other SQL pods using the PodNodeDialer instead of the NodeDialer that
handles SQL <-> KV communication in multi-tenant mode.

Fixes #74548

References #47900

Release note (sql change): Adds support for distributed import queries in
multi-tenant environments, which allows import queries to have improved
parallelism by utilizing all available SQL pods in the tenant.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76566)
<!-- Reviewable:end -->

rharding6373 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @stevendanna, and @yuzefovich)

___
*[pkg/ccl/changefeedccl/changefeeddist/distflow.go, line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mvvof9Q4OpkFcxXgSre-r2-50:-MvyMPR_5QeFKKe9K9Hu:b-gj19ft) ([raw file](https://github.com/cockroachdb/cockroach/blob/4014ce6978751ef9df8bedce14e4d7a85ba2478c/pkg/ccl/changefeedccl/changefeeddist/distflow.go#L50)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

why would we want to disable tenant distribution?
</blockquote></details>

It wasn't enabled before -- `distribute = ForSystemTenant()` meant that it was only distributed for the non-tenant case (not that it mattered, because `NewPlanningCtx()` also overrode distribution to disable it in the tenant case). The biggest reason not to enable it right now is we haven't tested tenant distribution with changefeeds yet. It might be interesting to turn tenant distribution on for some of these operations and see what breaks, but I haven't checked tests for use cases like changefeeds to see if there's already coverage for multi-pod tenants -- is there?


<!-- Sent from Reviewable.io -->

miretskiy commented:
I see -- I think we should enable it, but probably safer if we first expand
our test coverage.
Filed https://github.com/cockroachdb/cockroach/issues/76597

On Tue, Feb 15, 2022 at 12:24 PM Rachael Harding ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> *Reviewable <https://reviewable.io/reviews/cockroachdb/cockroach/76566>*
> status: [image: :shipit:] complete! 0 of 0 LGTMs obtained (waiting on
> @miretskiy <https://github.com/miretskiy>, @RaduBerinde
> <https://github.com/RaduBerinde>, @stevendanna
> <https://github.com/stevendanna>, and @yuzefovich
> <https://github.com/yuzefovich>)
>
> ------------------------------
>
> *pkg/ccl/changefeedccl/changefeeddist/distflow.go, line 50 at r2
> <https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mvvof9Q4OpkFcxXgSre-r2-50:-MvyMPR_5QeFKKe9K9Hu:b-gj19ft>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/4014ce6978751ef9df8bedce14e4d7a85ba2478c/pkg/ccl/changefeedccl/changefeeddist/distflow.go#L50>):*
> *Previously, miretskiy (Yevgeniy Miretskiy) wrote…*
>
> why would we want to disable tenant distribution?
>
> It wasn't enabled before -- distribute = ForSystemTenant() meant that it
> was only distributed for the non-tenant case (not that it mattered, because
> NewPlanningCtx() also overrode distribution to disable it in the tenant
> case). The biggest reason not to enable it right now is we haven't tested
> tenant distribution with changefeeds yet. It might be interesting to turn
> tenant distribution on for some of these operations and see what breaks,
> but I haven't checked tests for use cases like changefeeds to see if
> there's already coverage for multi-pod tenants -- is there?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/76566#pullrequestreview-883366880>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANA4FVGH63CJGUIYTB5S6KTU3KD5LANCNFSM5ONN7N6Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

RaduBerinde reviewed (commented):
Nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-:-MvyxJhW8aTs7aw6qkGA:bdtx8r6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @rharding6373, @stevendanna, and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 4004 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyuEmI3qhdGIFX1wZO:-MvyuEmI3qhdGIFX1wZP:b-ust2wy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_physical_planner.go#L4004)):*
> ```Go
> ) *PlanningCtx {
> 	// Tenants can only distribute plans with tenantDistributionEnabled.
> 	distribute = distribute && (tenantDistributionEnabled || evalCtx.Codec.ForSystemTenant())
> ```

I don't think we need both flags. We should either let the caller pass  either `true`, `false`, or `codec.ForSystemTenant()` for the `distribute` flag as necessary, or define a tri-state enum `Distribute, DistributeOnlySystemTenant, DontDistribute`
___
*[pkg/sql/distsql_plan_bulk.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyveYIDAG5ZlvVYgbm:-MvyveYIDAG5ZlvVYgbn:b-p6kwub) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L23)):*
> ```Go
> 
> type distSQLTenantPlanner interface {
> 	SetupAllNodesPlanning(
> ```

[nit] add a comment explaining what the method does
___
*[pkg/sql/distsql_plan_bulk.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyuforECXAHt6zs8fi:-MvyuforECXAHt6zs8fj:b-jv8x96) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L28)):*
> ```Go
> }
> 
> type singleTenantPlanner struct {
> ```

"single tenant" is not a great name, since you can be the system tenant in a multi-tenant configuration.
These should be `systemTenantPlanner` and `nonSystemTenantPlanner` (or just `tenantPlanner`). 
___
*[pkg/sql/distsql_plan_bulk.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mvyv6HqEN55U8ZCgRFN:-Mvyv6HqEN55U8ZCgRFO:bin6kp8) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L41)):*
> ```Go
> 	ctx context.Context, evalCtx *extendedEvalContext, execCfg *ExecutorConfig,
> ) (*PlanningCtx, []base.SQLInstanceID, error) {
> 	return dsp.distSQLTenantPlanner.SetupAllNodesPlanning(ctx, evalCtx, execCfg)
> ```

This is a bit mind-bending.. The tenant planner embeds a `*DistSQLPLanner` and the `DistSQLPlanner` contains the tenant planner as a field.

Maybe for now we should just have two internal `setupAllNodesPlanningSystem` and `setupAllNodesPlanningTenant` and have `SetupAllNodesPlanning` call the right one.

Otherwise, we should make a decision as to which interface is "above" the other and disentangle them. If we want the new interface to be "above" then it would be public and it would implement the public `SetupAllNodesPlanning` call (and could internally use the `DistSQLPlanner`). If it is "below", then it would be used by `DistSQLPlanner.SetupAllNodesPlanning` but it wouldn't use a `DistSQLPlanner` directly. It would be initialized with whatever necessary information and return the instance IDs (with `DistsqlPlanner.SetupAllNodesPlanning()` setting up the planning context separately). The system-tenant implementation of this interface would also contain `CheckInstanceHealthAndVersion` and other code that is system-tenant-only.
___
*[pkg/sql/distsql_plan_bulk.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mvyvdup8S7U4pKHbbDv:-Mvyvdup8S7U4pKHbbDw:b-d02fwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L44)):*
> ```Go
> }
> 
> func (dsp *singleTenantPlanner) SetupAllNodesPlanning(
> ```

[nit] add a comment mentioning this is part of the `distSQLTenantPlanner` interface.
___
*[pkg/sql/distsql_plan_bulk.go, line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyvY42DEMgNBHnkRlU:-MvyvY42DEMgNBHnkRlV:b-d02fwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L80)):*
> ```Go
> }
> 
> func (dsp *multiTenantPlanner) SetupAllNodesPlanning(
> ```

[nit] add a comment mentioning this is part of the `distSQLTenantPlanner` interface.
___
*[pkg/ccl/importccl/import_stmt_test.go, line 6944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyxTT2CB5uOsqvP6Ru:-MvyxTT2CB5uOsqvP6Rv:b-56uzsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4014ce6978751ef9df8bedce14e4d7a85ba2478c/pkg/ccl/importccl/import_stmt_test.go#L6944)):*
> ```Go
> 
> 	t1.Exec(t, createStmt)
> 	t1.Exec(t, importStmt, userfileURI, userfile2URI)
> ```

[nit] I wonder if there is a way to check that we're using two nodes.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice!

> Reviewed 19 of 19 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-:-MvzW7w802tAqnubjCvQ:b-83sjoh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, and @stevendanna)

___
*[pkg/ccl/changefeedccl/changefeeddist/distflow.go, line 50 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvzTpvC7Zn3UTIIL6P7:-MvzTpvC7Zn3UTIIL6P8:bhgk6o6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/ccl/changefeedccl/changefeeddist/distflow.go#L50)):*
> ```Go
> 	evalCtx := execCtx.ExtendedEvalContext()
> 	planCtx := dsp.NewPlanningCtx(ctx, evalCtx, nil /* planner */, noTxn,
> 		execCtx.ExecCfg().Codec.ForSystemTenant(), false /* tenantDistributionEnabled */)
> ```

nit: it'd be nice to keep the previous inlined comment (i.e. `/* distribute */`).
___
*[pkg/sql/distsql_physical_planner.go, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvzU98U6L8U9sKNXkH7:-MvzU98U6L8U9sKNXkH8:briwwfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_physical_planner.go#L130)):*
> ```Go
> 
> 	// sqlInstanceProvider has information about SQL instances in a multi-tenant
> 	// environment
> ```

nit: missing periods.
___
*[pkg/sql/distsql_physical_planner.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvzUD7o8DtpNKDZ_kBB:-MvzUD7o8DtpNKDZ_kBC:bk9jxz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_physical_planner.go#L133)):*
> ```Go
> 	sqlInstanceProvider sqlinstance.Provider
> 
> 	// bulk IO interface
> ```

nit: maybe add a few more words in the comment?
___
*[pkg/sql/distsql_physical_planner.go, line 4004 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyuEmI3qhdGIFX1wZO:-MvzUb9yEkNORiKAhQCw:ble5git) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_physical_planner.go#L4004)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think we need both flags. We should either let the caller pass  either `true`, `false`, or `codec.ForSystemTenant()` for the `distribute` flag as necessary, or define a tri-state enum `Distribute, DistributeOnlySystemTenant, DontDistribute`
</blockquote></details>

I like the idea of a tri-state enum - this way each caller has to explicitly think whether the distribution of its operation is supported in a multi-tenant setting.
___
*[pkg/sql/distsql_plan_bulk.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mvyv6HqEN55U8ZCgRFN:-MvzVLfTBVRKj0HMY7eb:b-3do0tx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L41)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is a bit mind-bending.. The tenant planner embeds a `*DistSQLPLanner` and the `DistSQLPlanner` contains the tenant planner as a field.

Maybe for now we should just have two internal `setupAllNodesPlanningSystem` and `setupAllNodesPlanningTenant` and have `SetupAllNodesPlanning` call the right one.

Otherwise, we should make a decision as to which interface is "above" the other and disentangle them. If we want the new interface to be "above" then it would be public and it would implement the public `SetupAllNodesPlanning` call (and could internally use the `DistSQLPlanner`). If it is "below", then it would be used by `DistSQLPlanner.SetupAllNodesPlanning` but it wouldn't use a `DistSQLPlanner` directly. It would be initialized with whatever necessary information and return the instance IDs (with `DistsqlPlanner.SetupAllNodesPlanning()` setting up the planning context separately). The system-tenant implementation of this interface would also contain `CheckInstanceHealthAndVersion` and other code that is system-tenant-only.
</blockquote></details>

+1 on the idea of `setupAllNodesPlanningSystem` and `setupAllNodesPlanningTenant`.
___
*[pkg/sql/distsql_running.go, line 334 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvzVXQaB0hllsgFsVap:-MvzVXQaB0hllsgFsVaq:b-ra7e4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3ac48dbccf121bdf7921de97d8b9619ba7693109/pkg/sql/distsql_running.go#L334)):*
> ```Go
> 			// NB: we wouldn't hit this in practice but if we did the actual
> 			// error would be opaque.
> 			return nil, nil, nil, errorutil.UnsupportedWithMultiTenancy(47900)
> ```

nit: maybe add a link to this issue in the PR description.
___
*[pkg/ccl/importccl/import_stmt_test.go, line 6944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyxTT2CB5uOsqvP6Ru:-MvzVzON7JxB839gHdZ0:bp7wd74) ([raw file](https://github.com/cockroachdb/cockroach/blob/4014ce6978751ef9df8bedce14e4d7a85ba2478c/pkg/ccl/importccl/import_stmt_test.go#L6944)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] I wonder if there is a way to check that we're using two nodes.
</blockquote></details>

I agree that it seems important for the test. Maybe we can check the trace?


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (commented):
TFTRs! I added multi-pod backup/restore tests to the second commit, but still need to figure out in-test verification of bulk op distribution.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-:-Mw3k2sV4HoOi-ef1e-1:b9bu2up)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, @stevendanna, and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvzUD7o8DtpNKDZ_kBB:-MvzZTcL-7Lkajl6m3yg:b-ltzjca) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_physical_planner.go#L133)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: maybe add a few more words in the comment?
</blockquote></details>

Removed in refactor.
___
*[pkg/sql/distsql_physical_planner.go, line 4004 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyuEmI3qhdGIFX1wZO:-Mw3bmGB1BiU4_-NOBIS:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_physical_planner.go#L4004)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I like the idea of a tri-state enum - this way each caller has to explicitly think whether the distribution of its operation is supported in a multi-tenant setting.
</blockquote></details>

Done
___
*[pkg/sql/distsql_plan_bulk.go, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyveYIDAG5ZlvVYgbm:-MvzUAXr183xeQGWipJA:benjlpu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L23)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add a comment explaining what the method does
</blockquote></details>

No longer applicable.
___
*[pkg/sql/distsql_plan_bulk.go, line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyuforECXAHt6zs8fi:-MvzUDBV15P05skFVJOm:becav8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"single tenant" is not a great name, since you can be the system tenant in a multi-tenant configuration.
These should be `systemTenantPlanner` and `nonSystemTenantPlanner` (or just `tenantPlanner`). 
</blockquote></details>

Renamed the functions below.
___
*[pkg/sql/distsql_plan_bulk.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mvyv6HqEN55U8ZCgRFN:-MvzTrK93PpPGm-rGgiS:bl62p4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L41)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

+1 on the idea of `setupAllNodesPlanningSystem` and `setupAllNodesPlanningTenant`.
</blockquote></details>

I removed the confusing interface for now. Thanks for the suggestion!
___
*[pkg/sql/distsql_plan_bulk.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mvyvdup8S7U4pKHbbDv:-MvzU3u78JoC09Jw1xU_:benjlpu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L44)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add a comment mentioning this is part of the `distSQLTenantPlanner` interface.
</blockquote></details>

No longer applicable.
___
*[pkg/sql/distsql_plan_bulk.go, line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyvY42DEMgNBHnkRlU:-MvzUL35ANdT51HsI3RC:byubvuk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d431ec62aa2dc7b19394cae6b6a67bd535acc94/pkg/sql/distsql_plan_bulk.go#L80)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] add a comment mentioning this is part of the `distSQLTenantPlanner` interface.
</blockquote></details>

No longer applicable
___
*[pkg/ccl/backupccl/backup_test.go, line 7059 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mw3jj5I-m8Y-f-w0kZZ:-Mw3jj5I-m8Y-f-w0kZ_:b-lg11ug) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e0fed8f7dbb7046fcfc02e8e7349729acaed5e4/pkg/ccl/backupccl/backup_test.go#L7059)):*
> ```Go
> 		})
> 
> 		t.Run("system-backup", func(t *testing.T) {
> ```

Need to verify that the backup and restore are distributed in the test (I've only verified this manually via distsql plans)
___
*[pkg/ccl/importccl/import_stmt_test.go, line 6944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyxTT2CB5uOsqvP6Ru:-Mw-CHm20QDpLWq27Xfp:bxjz1rl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4014ce6978751ef9df8bedce14e4d7a85ba2478c/pkg/ccl/importccl/import_stmt_test.go#L6944)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I agree that it seems important for the test. Maybe we can check the trace?
</blockquote></details>

Hm, although when I log the plan diagram internally it looks right: https://cockroachdb.github.io/distsqlplan/decode.html#eJyckDFPwzAQhXd-xs2micPmmSVLQawog2tfq0huzvjOCCnKf0dxhpIqINrx3t3z9_xG4I8ABkDBQB739owM5h00KGigUxATOWSmNMtjOWr9F5haQT_ELLPcKXCUEMwI0ktAMPCG1rfnSEmerdiqBgUexfahvJ4Z07EPaKrK49HmIP6wi_kQerdLRFIJsuwcf0I3KaAsFxCLPSEYPalfwlwy5IGSx4R-Be-mjbh7eqRYNVeH2-hmhdY39aDv6KG5s4jNNBxpYPzXP-u5KPQnXFplysnhayJXMMv4UnxF8MiybPUytENZlYA_zfpP89PKXF-bm5vI3fTwHQAA__-3d_dH

Tracing doesn't indicate anything happening on Node 2...

Still a work in progress.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, @stevendanna, and @yuzefovich)

___
*[pkg/ccl/importccl/import_stmt_test.go, line 6944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyxTT2CB5uOsqvP6Ru:-Mw8yk-62GcMp1YJFofm:b-cs9vrb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4014ce6978751ef9df8bedce14e4d7a85ba2478c/pkg/ccl/importccl/import_stmt_test.go#L6944)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Hm, although when I log the plan diagram internally it looks right: https://cockroachdb.github.io/distsqlplan/decode.html#eJyckDFPwzAQhXd-xs2micPmmSVLQawog2tfq0huzvjOCCnKf0dxhpIqINrx3t3z9_xG4I8ABkDBQB739owM5h00KGigUxATOWSmNMtjOWr9F5haQT_ELLPcKXCUEMwI0ktAMPCG1rfnSEmerdiqBgUexfahvJ4Z07EPaKrK49HmIP6wi_kQerdLRFIJsuwcf0I3KaAsFxCLPSEYPalfwlwy5IGSx4R-Be-mjbh7eqRYNVeH2-hmhdY39aDv6KG5s4jNNBxpYPzXP-u5KPQnXFplysnhayJXMMv4UnxF8MiybPUytENZlYA_zfpP89PKXF-bm5vI3fTwHQAA__-3d_dH

Tracing doesn't indicate anything happening on Node 2...

Still a work in progress.
</blockquote></details>

One idea: make a table with a column who's default value is `row int primary key, instance int default crdb_internal.node_id()` (or add a companion builtin crdb_internal.instance_id?) then IMPORT INTO that table two files that just have rows "1", "2", "3" and "4", "5", "6". 

Now right now this IMPORT will fail because the crdb_internal.node_id builtin is too volatile. But we could add a `TestingAddImportBuiltin(name, *CustomFunc)` to pkg/sql/row/expr_walker.go to inject an entry into the `supportedImportFuncOverrides` map, to allow calling it in the test. 

then we just inspect the populated instance IDs in the table e.g.`select r from [select count(*) as r, instance_id from t group by instance_id]` should look like 3, 3?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, @stevendanna, and @yuzefovich)

___
*[pkg/ccl/importccl/import_processor.go, line 177 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MwXhp1C4c1CkCTLFQd1:-MwXhp1C4c1CkCTLFQd2:btshm5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f1b10b842adf5c5099ff8f1aae7ad01624b880/pkg/ccl/importccl/import_processor.go#L177)):*
> ```Go
> // Start is part of the RowSource interface.
> func (idp *readImportDataProcessor) Start(ctx context.Context) {
> 	log.Infof(ctx, "starting read import")
> ```

These calls need to happen after `StartInternal` call below. There, we derive a tracing span with a new context, so in order for these messages appear in the trace, we need to use the new context (the third commit had the right version).
___
*[pkg/ccl/importccl/import_stmt_test.go, line 6951 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MwXi2lH29g8oRubfhaI:-MwXi2lIBTflsyZV8smg:bwgdk07) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f1b10b842adf5c5099ff8f1aae7ad01624b880/pkg/ccl/importccl/import_stmt_test.go#L6951)):*
> ```Go
> 	t1.Exec(t, "SET tracing = off")
> 	t2.Exec(t, "SET tracing = off")
> 	rows := t1.Query(t, "SELECT concat(location, ' ', tag, ' ', span::string, ' ', operation, ' ', message, ' ', timestamp::string) FROM [SHOW TRACE FOR SESSION]")
> ```

I recently encountered a case where `SHOW TRACE FOR SESSION` also didn't include some of the events that I knew for sure occurred. I talked with Andrei about it, he suggested that the session tracing might be broken and he wants to get rid off it.

First, I'd use `EXPLAIN ANALYZE (DEBUG)` on the query for which you want to see the trace to make sure that events do get into the trace there. Next, I'd refactor the test to use `WithStatementTrace` testing knob (take a look at `TestLargeKeys` in `kvstreamer/large_keys_test.go` for an example).


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
r1 + r2 LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-:-MwCE7i_7knOgUJBGb5O:b-ojkav1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, @stevendanna, and @yuzefovich)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
I meant first two commits LGTM

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-:-MwYanlzAtngOKW8816-:bda79k9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, @stevendanna, and @yuzefovich)



<!-- Sent from Reviewable.io -->

rharding6373 committed:
importccl, backupccl: add import and restore tenant tests with multiple SQL pods

These tests validate that imports and restores can be
distributed across multiple SQL pods in a single tenant.

Release note: None

rharding6373 reviewed (commented):
TFTRs! RFAL

I've been unable to verify in-test that the bulk ops are distributed, so I left TODOs in the interest of getting this in before code freeze. I have done manual verification using logs, and here is e.g. a distributed restore plan diagram: https://cockroachdb.github.io/distsqlplan/decode.html#eJy0ksFuszAQhO__U0Rz9q-ASVKJW6Veckmr5lhxoPYqRaU2tY3UKuLdKyBSAk1NUujRLN_uzOzuYd9zxACD0pI26RtZxE8IwcDBECFhKIwWZK02dWnf_LiWH4gDhkwVpas_JwxCG0K8h8tcToixLfLM3Sq5FalzZLYFiXkABkkuzfJmDJ-Jl1K9WiQVgy7dodmxx_PnzKRqRx0sqRIG69IdIQ4r9oOiY5NSaSPJkPzepa_5kazThu5SlzZ6ox5yqvIogk8rYqP_62K-umh01BkdXr8R3t1IOPFGwqk3svzFRvj1sYR_e6h86lgWIw_1_IhCK9u3dL5zUIskuaPWkdWlEfRgtGjGtM_7hmtuQpJ1bfXwWKu2VAu8HF6NgRd-OOzDwSnM_TD3wjcdOOjD0ZjA_PBAYH54ILDFmMCWYzz74QHPfnjA8-oqz0n17ysAAP__L_l-2w==

Backups are currently not distributed because they rely on PartitionSpans, which I will fix for the tenant case in a follow-up PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-:-Mwcx67b6Ll7lCGgDeVf:b-dngjwo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, @stevendanna, and @yuzefovich)

___
*[pkg/ccl/importccl/import_processor.go, line 177 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MwXhp1C4c1CkCTLFQd1:-Mwchmt-2Nc8mKvXKK5m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f1b10b842adf5c5099ff8f1aae7ad01624b880/pkg/ccl/importccl/import_processor.go#L177)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

These calls need to happen after `StartInternal` call below. There, we derive a tracing span with a new context, so in order for these messages appear in the trace, we need to use the new context (the third commit had the right version).
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/import_stmt_test.go, line 6944 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MvyxTT2CB5uOsqvP6Ru:-MwchuZ85LA5Ntri8fxP:bhbdt01) ([raw file](https://github.com/cockroachdb/cockroach/blob/4014ce6978751ef9df8bedce14e4d7a85ba2478c/pkg/ccl/importccl/import_stmt_test.go#L6944)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

One idea: make a table with a column who's default value is `row int primary key, instance int default crdb_internal.node_id()` (or add a companion builtin crdb_internal.instance_id?) then IMPORT INTO that table two files that just have rows "1", "2", "3" and "4", "5", "6". 

Now right now this IMPORT will fail because the crdb_internal.node_id builtin is too volatile. But we could add a `TestingAddImportBuiltin(name, *CustomFunc)` to pkg/sql/row/expr_walker.go to inject an entry into the `supportedImportFuncOverrides` map, to allow calling it in the test. 

then we just inspect the populated instance IDs in the table e.g.`select r from [select count(*) as r, instance_id from t group by instance_id]` should look like 3, 3?
</blockquote></details>

The consensus from offline discussions has been to leave verification of distribution in the test as a TODO for now since tracing is broken for import and fixing it would be a bit of effort, as would David's suggestion, and EXPLAIN doesn't work well for bulk ops. I think using tracing would be ideal since it would translate to backup/restore well, as well.
___
*[pkg/ccl/importccl/import_stmt_test.go, line 6951 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-MwXi2lH29g8oRubfhaI:-Mwcx1P7CNHzzFaJxHmh:b572c1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7f1b10b842adf5c5099ff8f1aae7ad01624b880/pkg/ccl/importccl/import_stmt_test.go#L6951)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I recently encountered a case where `SHOW TRACE FOR SESSION` also didn't include some of the events that I knew for sure occurred. I talked with Andrei about it, he suggested that the session tracing might be broken and he wants to get rid off it.

First, I'd use `EXPLAIN ANALYZE (DEBUG)` on the query for which you want to see the trace to make sure that events do get into the trace there. Next, I'd refactor the test to use `WithStatementTrace` testing knob (take a look at `TestLargeKeys` in `kvstreamer/large_keys_test.go` for an example).
</blockquote></details>

Thanks for the tips.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 5 of 19 files at r3, 7 of 16 files at r5, 15 of 15 files at r9, 6 of 6 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, and @stevendanna)

___
*[-- commits, line 29 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mwe44Hw9R38wjfei4Gv):*
nit: `s/Theses/These/`.
___
*[pkg/sql/apply_join.go, line 309 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mwe2zVq6hG4aW5vCi2k:-Mwe2zVrDI4jSPGfvkn5:bo30984) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43f7a262eea9b85d3a7f8134318c1f4ecc9e5d0/pkg/sql/apply_join.go#L309)):*
> ```Go
> 		params.ctx, &plannerCopy, plannerCopy.execCfg.NodeID, plannerCopy.SessionData().DistSQLMode, plan.main,
> 	)
> 	var distributeType DistributionType
> ```

nit: slightly more concise version which seems to be more common in our codebase:
```
distributeType := DistributionTypeNode
if distributePlan.WillDistribute() {
  distributeType = DistributionTypeSystemTenantOnly
}
```
___
*[pkg/sql/importer/import_processor.go, line 185 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mwe4I-e6VShltsfhodi:-Mwe4I-e6VShltsfhodj:b-fvu89r) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6636f915cd1ff90243333c5f44dafaf637b60/pkg/sql/importer/import_processor.go#L185)):*
> ```Go
> 	ctx = idp.StartInternal(ctx, readImportDataProcessorName)
> 	log.Infof(ctx, "starting read import")
> 	log.VEvent(ctx, 2, "starting read import")
> ```

These seem redundant, no?


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, @stevendanna, and @yuzefovich)

___
*[pkg/sql/apply_join.go, line 309 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mwe2zVq6hG4aW5vCi2k:-MweQNhQ2ITNpem1wJkX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c43f7a262eea9b85d3a7f8134318c1f4ecc9e5d0/pkg/sql/apply_join.go#L309)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: slightly more concise version which seems to be more common in our codebase:
```
distributeType := DistributionTypeNode
if distributePlan.WillDistribute() {
  distributeType = DistributionTypeSystemTenantOnly
}
```
</blockquote></details>

Done.
___
*[pkg/sql/importer/import_processor.go, line 185 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mwe4I-e6VShltsfhodi:-MweQUxFEHvTRkdUh4IN:bp89ymj) ([raw file](https://github.com/cockroachdb/cockroach/blob/90e6636f915cd1ff90243333c5f44dafaf637b60/pkg/sql/importer/import_processor.go#L185)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

These seem redundant, no?
</blockquote></details>

A bit, I suppose. Removed.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r11, 3 of 4 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-:-MwgnKWdAAIZZF8RpCgp:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy, @RaduBerinde, @rharding6373, and @stevendanna)

___
*[pkg/sql/distsql_plan_ctas.go, line 34 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mwgn8-Z1BI8bRm2p2cS:-Mwgn8-Z1BI8bRm2p2cT:bmvl5v8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e978737454a50ce4600d350fc71918ed3dc1171/pkg/sql/distsql_plan_ctas.go#L34)):*
> ```Go
> 	recv *DistSQLReceiver,
> ) {
> 	var distribute DistributionType
> ```

nit: there are a couple more places we could make things more concise and consistent.


<!-- Sent from Reviewable.io -->

rharding6373 reviewed (commented):
TFTRs and all the help getting this together! Refactored the backup/restore test due to some resource limitations and disable race testing to avoid flakiness due to the test size. It looks like remaining failures have been known flakes.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-:-MwhQNvfBnNZg2Qiia1r:bl9xx7g)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @miretskiy, @RaduBerinde, @rharding6373, @stevendanna, and @yuzefovich)

___
*[pkg/sql/distsql_plan_ctas.go, line 34 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/76566#-Mwgn8-Z1BI8bRm2p2cS:-MwhTF0C0FX4NtNaf7nz:b-z4od6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e978737454a50ce4600d350fc71918ed3dc1171/pkg/sql/distsql_plan_ctas.go#L34)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: there are a couple more places we could make things more concise and consistent.
</blockquote></details>

Thanks, looks like I missed a bunch...


<!-- Sent from Reviewable.io -->

rharding6373 commented:
Failed due to a flaky test.

bors r+

rharding6373 commented:
bors r+
